File: src/main/java/com/mbayou/kick4k/api/ApiResponseDeserializer.java
package com.mbayou.kick4k.api;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.ObjectCodec;
import com.fasterxml.jackson.databind.BeanProperty;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.deser.ContextualDeserializer;

import java.io.IOException;

public class ApiResponseDeserializer<T> extends JsonDeserializer<ApiResponse<T>> implements ContextualDeserializer {
    private JavaType valueType;

    public ApiResponseDeserializer() {
    }

    public ApiResponseDeserializer(JavaType valueType) {
        this.valueType = valueType;
    }

    @Override
    public ApiResponse<T> deserialize(JsonParser parser, DeserializationContext ctx) throws IOException {
        ObjectCodec codec = parser.getCodec();
        JsonNode root = codec.readTree(parser);

        JsonNode dataNode = root.get("data");
        T data;

        if (dataNode == null || dataNode.isNull() || (dataNode.isObject() && dataNode.isEmpty())) {
            data = null;
        } else {
            ObjectMapper mapper = (ObjectMapper) codec;
            data = mapper.readerFor(valueType).readValue(dataNode.traverse(mapper));
        }

        String message = root.has("message") ? root.get("message").asText(null) : null;
        return new ApiResponse<>(data, message);
    }

    @Override
    public JsonDeserializer<?> createContextual(DeserializationContext ctx, BeanProperty property) {
        JavaType contextualType = (property != null) ? property.getType() : ctx.getContextualType();

        if (contextualType == null || !contextualType.hasGenericTypes()) {
            return this;
        }

        JavaType contentType = contextualType.containedType(0);
        return new ApiResponseDeserializer<>(contentType);
    }
}

File: src/main/kotlin/com/mbayou/kick4k/api/ApiResponseDeserializer.kt
package com.mbayou.kick4k.api

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.BeanProperty
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.deser.ContextualDeserializer
import java.io.IOException

class ApiResponseDeserializer<T>(private val valueType: JavaType? = null) :
    JsonDeserializer<ApiResponse<T>>(),
    ContextualDeserializer {

    override fun deserialize(parser: JsonParser, context: DeserializationContext): ApiResponse<T> {
        val codec = parser.codec
        val root = codec.readTree(parser) as JsonNode

        val dataNode = root.get("data")
        val data: T? = if (dataNode == null || dataNode.isNull || (dataNode.isObject && dataNode.isEmpty)) {
            null
        } else {
            val mapper = codec as ObjectMapper
            val type = valueType ?: mapper.typeFactory.constructType(Any::class.java)
            mapper.readerFor(type).readValue(dataNode.traverse(mapper))
        }

        val message = if (root.has("message")) root.get("message").asText(null) else null
        return ApiResponse(data, message)
    }

    override fun createContextual(context: DeserializationContext, property: BeanProperty?): JsonDeserializer<*> {
        val contextualType = property?.type ?: context.contextualType
        if (contextualType == null || !contextualType.hasGenericTypes()) {
            return this
        }

        val contentType = contextualType.containedType(0)
        return ApiResponseDeserializer<Any>(contentType)
    }
}
